applications/ShapeOptimizationApplication/tests/sensitivity_analysis_test/run_sensitivity_verification_in_geometry_space.py

```diff
@@ -0,0 +1,120 @@
+    with open(results_filename,'a') as open_file:
+        open_file.write("-----------------------------------------------------------------------\n")
+        open_file.write("time\t\t\t\t= "+str(time.ctime())+"\n")
+        open_file.write("move_node_id\t\t= "+str(move_node_id)+"\n")
+        open_file.write("reference_value\t\t= "+str(reference_value)+"\n")
+        open_file.write("reference_gradient\t= "+str(reference_gradient[move_node_id])+"\n")
+        open_file.write("---\n")
+        open_file.write("step,\tfd_gradient_x,\tfd_gradient_y,\tfd_gradient_z,\ttime\n")
+
+    for itr in finite_difference_levels:
+        current_delta = 1*10**(-itr)
+
+        for dim_itr in range(domain_size):
+
+            print("\n\n##########################################################################")
+            print(">> Starting finite difference level ",itr, " dimension ",dim_itr+1, "for node ", str(move_node_id))
+            print("##########################################################################\n\n")
+
+            # Initialize new analysis
+            model = KM.Model()
+            response = structural_response_function_factory.CreateResponseFunction(kratos_response_settings["response_type"].GetString(), kratos_response_settings, model)
+            response.Initialize()
+
+            # Perturb
+            model_part = model.GetModelPart(model_part_name)
+            if dim_itr == 0:
+                model_part.Nodes[move_node_id].X0 += current_delta
+                model_part.Nodes[move_node_id].X += current_delta
+            elif dim_itr == 1:
+                model_part.Nodes[move_node_id].Y0 += current_delta
+                model_part.Nodes[move_node_id].Y += current_delta
+            elif dim_itr == 2:
+                model_part.Nodes[move_node_id].Z0 += current_delta
+                model_part.Nodes[move_node_id].Z += current_delta
+
+            # Perform analysis
+            response.InitializeSolutionStep()
+            response.CalculateValue()
+            response.FinalizeSolutionStep()
+            response.Finalize()
+            current_value = response.GetValue()
+
+            # Evaluate fd gradients
+            if dim_itr == 0:
+                fd_gradient_x = (current_value-reference_value)/current_delta
+            elif dim_itr == 1:
+                fd_gradient_y = (current_value-reference_value)/current_delta
+            elif dim_itr == 2:
+                fd_gradient_z = (current_value-reference_value)/current_delta
+
+        # Write results
+        with open(results_filename,'a') as open_file:
+            line_to_write = '%.0E' % Decimal(str(current_delta)) + ",\t"
+            line_to_write += '%.6E' % Decimal(str(fd_gradient_x)) + ",\t"
+            line_to_write += '%.6E' % Decimal(str(fd_gradient_y)) + ",\t"
+            line_to_write += '%.6E' % Decimal(str(fd_gradient_z)) + ",\t"
+            line_to_write += str(time.ctime()) + "\n"
+            open_file.write(line_to_write)
```

Comment:
what is the difference between this and run_semi_analytic_step_size_verification.py ?

Reply:
In "run_semi_analytic_step_size_verification.py", basically a step-size study is done to see if for different step-sizes similar gradients are obtained by the semi-analytic approach. Whereas, in the "run_sensitivity_verification_in_geometry_space.py", the entire gradient for selected nodes are computed by finite differencing and varying the grid coordinates (no semi-analytic approach here). Theoretically, the latter has to match the former.